[Jonoans] iP

Jonoans opened the pull request.

Jonoans committed:
Completed Level-0

Jonoans committed:
Added echo and bye command

Jonoans committed:
Added task addition and listing

Jonoans committed:
Add task marking and unmarking features

Jonoans committed:
Refactor done variable to isDone - conform to coding standard

Jonoans committed:
Check for empty inputs

Jonoans committed:
Refactor code to comply to coding standard

Jonoans committed:
Refactor final task variable to all caps

Jonoans committed:
Refactor ApplicationState
- Move implementation details for tasks state into task class

Jonoans committed:
Add Command class to parse commands

Jonoans committed:
Remove unused imports and variable assignment

Jonoans committed:
Add task types deadline, event and todo

Jonoans committed:
Change print message when adding tasks

Jonoans committed:
Add testcases for UI test

J0shuaLeong commented on src/main/java/ProgramConstants.java:
```diff
@@ -0,0 +1,11 @@
+public final class ProgramConstants {
+    public static final String BOT_NAME = "DumbGPT";
+    public static final String SEPARATOR = "____________________________________________________________";
+
```
Nice to see constants follows the standard

J0shuaLeong commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,160 @@
+            return taskNumber;
+        }
+
+        if (taskNumber <= 0) return -1;
```
According to the coding standard, even if there is only one statement in an if block, should be within {} and on a new line.

J0shuaLeong commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,160 @@
+        ended = true;
+        return null;
     }
 }
```
Very neat and well written code!

Jonoans committed:
Fix coding standard violation

DavinciDelta commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,160 @@
+        return SCANNER.nextLine();
+    }
+
+    public static Void addTaskFlow(Command command) {
```
i like the use of switch case to avoid deep nesting!

Jonoans committed:
Add missing space

DavinciDelta commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,160 @@
+    private final static Scanner SCANNER = new Scanner(System.in);
+    private final static HashMap<String, Function<Command, Void>> COMMANDS = new HashMap<>();
+
     public static void main(String[] args) {
```
I feel that this method length seems to be unnecessarily long, perhaps you could try to shorten it?

Jonoans replied:
I am assuming you are referencing line 9, if so, I think it's a little difficult for me to shorten it

DavinciDelta reviewed (commented):
good code and great use of HashMap, this allows adding in future commands more conveniently!

Jonoans committed:
Re-organise code into packages

Jonoans committed:
Add new method printWrapped, accepting String array for printing

Jonoans committed:
Refactor code
- Rename variables in Command class for better comprehension
- Rename "states" in ApplicationState to "STATES"

Jonoans committed:
Remove integer literal

Jonoans committed:
Rename variables and pass tokens list directly to addOption in Command class

Jonoans committed:
Extract execution flow to method

Jonoans committed:
Implement each runnable command as a class on its own

Jonoans committed:
Provide feedback when invalid command is provided
- Move checking of empty command into ActionRegistry's execute method

Jonoans committed:
Ensure alias string is not blank

Jonoans committed:
Fix bug where unmark command marks tasks as done instead of marking as undone

Jonoans committed:
Make action class abstract

Jonoans committed:
Remove override annotation from direct child of Action class

Jonoans committed:
Store isEnded in boolean variable outside of STATES

Jonoans committed:
Update tests to include invalid inputs

Jonoans committed:
Merge branch 'branch-A-Packages'

Jonoans committed:
Add DukeException class
- All program errors will be thrown as DukeException errors
- Add handler code for DukeException error in ActionRegistry execute method

Jonoans committed:
Merge branch 'branch-Level-5'

Jonoans committed:
Move getTask function out of MarkTask class, rename variable in ProgramConstants

Jonoans committed:
Add task deletion command

Jonoans committed:
Update tests to include delete command

Jonoans committed:
Add save functionality for tasks

Jonoans committed:
Merge branch 'branch-Level-6'

Jonoans committed:
Merge branch 'branch-Level-7'

kohnh commented on src/main/java/com/gpt/dumpgpt/Duke.java:
```diff
@@ -0,0 +1,69 @@
+package com.gpt.dumpgpt;
+
+import com.gpt.dumpgpt.action.api.ActionRegistry;
+import com.gpt.dumpgpt.action.impl.*;
```
Try to explicitly list your imports

kohnh commented on src/main/java/com/gpt/dumpgpt/action/api/ActionRegistry.java:
```diff
@@ -0,0 +1,99 @@
+    private static ActionRegistry actionRegistry;
+    private final HashMap<String, Constructor<? extends Action>> ACTIONS = new HashMap<>();
+
+    private ActionRegistry() {
+    }
```
Try not to include constructors if it does nothing

kohnh commented on src/main/java/com/gpt/dumpgpt/action/api/ActionRegistry.java:
```diff
@@ -0,0 +1,99 @@
+        String[] aliases = action.getAliases();
+
+        boolean isVerbValid = isActionVerbValid(actionVerb),
+                isAliasesValid = isAliasesValid(aliases);
```
Try not to name your variables the same as your methods which might make it unclear

kohnh commented on src/main/java/com/gpt/dumpgpt/action/impl/AddTask.java:
```diff
@@ -0,0 +1,67 @@
+            return new Event(command.getArguments(), command.getOptions("from"), command.getOptions("to"));
+        case "todo":
+            return new Todo(command.getArguments());
+        }
```
Try to have a `default:` branch to handle other cases

kohnh commented on src/main/java/com/gpt/dumpgpt/action/impl/DeleteTask.java:
```diff
@@ -0,0 +1,50 @@
+                "I've deleted the following task:",
+                "\t" + task.toString(),
+                taskSummary
+        });
```
You can consider abstracting some parts of your code to not repeat yourself.

kohnh commented on src/main/java/com/gpt/dumpgpt/command/Command.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public void parseTokens() {
+        if (IS_EMPTY) {
```
Since you have created a getter, you can consider using it

kohnh commented on src/main/java/com/gpt/dumpgpt/shared/Serializer.java:
```diff
@@ -0,0 +1,224 @@
+        return ByteBuffer.wrap(value).getInt();
+    }
+
+    public void assignType(String objectType) {
```
You can consider using the prefix `set` for your setters and `get` for your getters to standardize

kohnh reviewed (commented):
Minor code quality and coding standards issues.

Jonoans committed:
Add JavaDocs

Jonoans committed:
Add OOP classes

Jonoans committed:
Merge branch 'branch-A-MoreOOP' into branch-A-JavaDoc

Jonoans committed:
Merge pull request #2 from Jonoans/branch-A-MoreOOP

A-MoreOOP

Jonoans committed:
Merge pull request #3 from Jonoans/branch-A-JavaDoc

A-JavaDoc

Jonoans committed:
Merge branch 'branch-A-MoreOOP' into branch-Level-9

Jonoans committed:
Merge branch 'branch-A-MoreOOP' into branch-Level-8

Jonoans committed:
Add support for datetime formatting

Jonoans committed:
Add find command

Jonoans committed:
Merge branch 'master' into branch-Level-8

Jonoans committed:
Merge pull request #5 from Jonoans/branch-Level-9

Level-9

Jonoans committed:
Merge remote-tracking branch 'origin/master' into branch-Level-8

Jonoans committed:
Merge pull request #4 from Jonoans/branch-Level-8

Level-8

Jonoans committed:
Add custom class for datetime

Jonoans committed:
Merge pull request #6 from Jonoans/branch-Level-8

Add custom class for datetime features

Jonoans committed:
Set separator for date and time to dash

Jonoans committed:
Merge pull request #7 from Jonoans/branch-Level-8

Change separator for date and time in printouts

Jonoans committed:
Merge branch 'master' into branch-A-JavaDoc

Jonoans committed:
Add docs Serializer, Storage and DukeDateTime class

Jonoans committed:
Merge pull request #8 from Jonoans/branch-A-JavaDoc

Add more javadoc

Jonoans committed:
Add userguide

Jonoans committed:
Add command summary and header link section

Jonoans committed:
Merge branch 'branch-A-UserGuide'

Jonoans committed:
Include default case

Jonoans committed:
Add theme for userguide

Jonoans committed:
Fix remote theme configuration

Jonoans committed:
Fix callout format

Jonoans committed:
Configure theme

Jonoans committed:
Minor changes to userguide

Jonoans committed:
Merge branch 'master' into branch-Level-8

Jonoans committed:
Ensure localDatetime passed to format is non-null

Jonoans committed:
Merge pull request #9 from Jonoans/branch-Level-8

Fix potential bug in format method of DukeDateTime
